libftl/charon_protocol.c

```diff
@@ -52,3 +60,61 @@ ftl_charon_response_code_t ftl_charon_read_response_code(const char * response_s
    /* Got an invalid or unknown response code */
    return FTL_CHARON_UNKNOWN;
  }
+
+unsigned char decode_hex_char(char c) {
+    if (c >= '0' && c <= '9') {
+        return c - '0';
+    }
+    if (c >= 'A' && c <= 'Z') {
+        return (c - 'A') + 10;
+    }
+    if (c >= 'a' && c <= 'z') {
+        return (c - 'a') + 10;
+    }
+
+    return 0;
+}
+
+int recv_all(int sock, char * buf, int buflen) {
+    int pos = 0;
+    while (pos == 0 || buf[pos - 1] != '\n') {
+        pos += recv(sock, buf + pos, buflen - pos, 0);
+        if (pos >= buflen) {
+            return pos;
+        }
+    }
+    return pos;
+}
+
+int ftl_charon_get_hmac(int sock, char * auth_key, char * dst) {
+    char buf[2048];
+    int string_len;
+    int response_code;
+
+    send(sock, "HMAC\r\n\r\n", 8, 0);
+    string_len = recv_all(sock, buf, 2048);
+
+    response_code = ftl_charon_read_response_code(buf);
```

Comment:
Add length checking after error checking recv_all to prevent reading uninitialized memory